// @hubspot/ui-extensions/dist/internal/hook-utils.js
import { createContext, useContext } from 'react';
const MocksContext = createContext(null);
/**
 * Creates a mock-aware hook function that can be used to mock the original hook function.
 * The mock-aware hook function will return the mocked hook function if a mock is found, otherwise it will return the original hook function.
 *
 * @param hookName The name of the hook to mock that corresponds to the key in the Mocks interface
 * @param originalHookFunction The original hook function to call if no mock is found
 * @returns The mocked hook function or the original hook function if no mock is found
 */
export const createMockAwareHook = (hookName, originalHookFunction) => {
    const useWrapper = (...args) => {
        const mocks = useContext(MocksContext);
        if (!mocks) {
            // If no mocks are provided, call the original hook function
            return originalHookFunction(...args);
        }
        // When a mock is provided by the testing utilities (via <MocksContextProvider>), return the mocked hook function
        const mockHook = mocks[hookName];
        if (!mockHook) {
            throw new Error(`Illegal State: Mock for hook ${hookName} not found.`);
        }
        // Call the mocked hook function with the same arguments as the original hook function and return the result
        return mockHook(...args);
    };
    return useWrapper;
};
/**
 * A React component that provides the Mocks context that can be used to provide mocks to the mock-aware hook functions.
 *
 * @param children The children to render.
 * @returns The children wrapped in the Mocks context provider.
 */
export const MocksContextProvider = MocksContext.Provider;
